src/tools/rbd_mirror/group_replayer/Replayer.cc

```diff
@@ -542,113 +570,126 @@ void Replayer<I>::validate_image_snaps_sync_complete(
+        auto mirror_ns = std::get_if<cls::rbd::MirrorSnapshotNamespace>(
+                                                &snap_info.snapshot_namespace);
+        if (!mirror_ns) {
+          continue;
+        }
+        // Makesure the image snapshot is COMPLETE
+        if (mirror_ns->group_snap_id == group_snap_id && mirror_ns->complete) {
+          image_snap_complete = true;
+          cls::rbd::ImageSnapshotSpec snap_spec;
+          snap_spec.pool = image.spec.pool_id;
+          snap_spec.image_id = image.spec.image_id;
+          snap_spec.snap_id = snap_info.id;
+          auto it = std::find_if(
+              itl->snaps.begin(), itl->snaps.end(),
+              [&snap_spec](const cls::rbd::ImageSnapshotSpec &s) {
+              return snap_spec.pool == s.pool && snap_spec.image_id == s.image_id;
+              });
+          if (it == itl->snaps.end()) {
+            image_snap_spec.push_back(snap_spec);
+          }
+          continue;
+        } else {
+          dout(10) << "remote group snap id: " << group_snap_id
+                   << ", local reflected in the image snap: "
+                   << mirror_ns->group_snap_id << dendl;
+        }
+      }
+      locker.unlock();
+
+      // In case if the image replayer is not yet updated with the limits, do it here
+      if (!image_snap_complete) {
+        set_image_replayer_limits(image.spec.image_id, &(*itr));
       }
+      image_snap_complete = false; // reset
     }
-    locker.unlock();
 
-    // In case if the image replayer is not yet updated with the limits, do it here
-    if (!image_snap_complete) {
-      set_image_replayer_limits(image.spec.image_id, &(*itr));
+    locker.lock();
+    for (auto itl = m_local_group_snaps.begin();
+         itl != m_local_group_snaps.end(); ++itl) {
+      if (itl->id == group_snap_id) {
+        break;
+      } else if (itl->state == cls::rbd::GROUP_SNAPSHOT_STATE_INCOMPLETE) {
+        // there would definitely be a regular group snap newly added, so lets wait for it.
+        return;
+      }
     }
-    image_snap_complete = false; // reset
-  }
 
-  if (itr->snaps.size() == image_snap_spec.size()) {
-    std::unique_lock locker{m_lock};
-    for (auto &spec : image_snap_spec) {
-      C_SaferCond *ctx = new C_SaferCond;
-      mirror_snapshot_complete(group_snap_id, &spec, ctx);
-      r = ctx->wait();
-      if (r < 0) {
-        m_retry_validate_snap = true;
+    if (!m_prune_snap_ids.empty()) {
+      // this would be regular group snap added for removal, so lets wait for it.
+      goto out;
+    } else if (itr->snaps.size() == image_snap_spec.size()) {
+      for (auto &spec : image_snap_spec) {
+        C_SaferCond *ctx = new C_SaferCond;
+        mirror_snapshot_complete(group_snap_id, &spec, ctx);
+        r = ctx->wait();
+        if (r < 0) {
+          m_retry_validate_snap = true;
+        }
       }
     }
-    locker.unlock();
   }
+
+out:
+  prune_unnecessary_mirror_group_snaps(locker);
```

Comment:
```
In load_local_group_snapshots(): 
for (auto &local_snap : m_local_group_snaps) {
...
  --> validate_image_snaps_sync_complete(local_snap.snap_id, locker)
      --> prune_unnecessary_mirror_group_snaps(locker)
          --> prune_group_snapshots(locker)
              --> prune_group_snapshot(prune_snap_id, locker)
                  -->   for (auto local_snap = m_local_group_snaps.begin();local_snap != m_local_group_snaps.end(); ++local_snap) {

...
                          local_snap =  m_local_group_snaps.erase(local_snap);
```
This could mess up the iterators in the outer loop. (Thanks for clarifying @idryomov )

Reply:
This is seen in the following scenario:
Test steps:

1. Create an empty group and create a few user snaps on it.
2. Mirror enable the group and and wait until it is synced.
3. Remove one of the user group snaps and create a mirror group snapshot

The secondary rbd-mirror daemon crashes